File: C/lcr168-chou-shu-lcof/main.go
package main

import (
	"fmt"
)

func nthUglyNumber(n int) int {
	buf := make([]int, n)
	buf[0] = 1
	i2, i3, i5 := 0, 0, 0
	for i := 1; i < n; i++ {
		v2, v3, v5 := buf[i2]*2, buf[i3]*3, buf[i5]*5
		min := v2
		if min > v3 {
			min = v3
		}
		if min > v5 {
			min = v5
		}
		buf[i] = min
		if min == v2 {
			i2++
		}
		if min == v3 {
			i3++
		}
		if min == v5 {
			i5++
		}
	}
	return buf[n-1]
}

func main() {
	fmt.Println(nthUglyNumber(1352))
}

File: C/lcr168-chou-shu-lcof/main.rs
struct Solution {}

impl Solution {
  pub fn nth_ugly_number(n: i32) -> i32 {
    let mut buf: Vec<i32> = vec![0; n as usize];
    buf[0] = 1;
    let mut i2: usize = 0;
    let mut i3: usize = 0;
    let mut i5: usize = 0;
    for i in 1..n {
      let v2 = buf[i2] * 2;
      let v3 = buf[i3] * 3;
      let v5 = buf[i5] * 5;
      let mut min = v2;
      if min > v3 {
        min = v3;
      }
      if min > v5 {
        min = v5;
      }
      buf[i as usize] = min;
      if min == v2 {
        i2 += 1;
      }
      if min == v3 {
        i3 += 1;
      }
      if min == v5 {
        i5 += 1;
      }
    }
    buf[n as usize - 1]
  }
}

fn main() {
  println!("{}", Solution::nth_ugly_number(1352));
}
